ml4gw/utils/slicing.py

```diff
@@ -13,6 +14,67 @@
 ]
 
 
+def unfold_windows(
+    x: torch.Tensor,
+    window_size: int,
+    stride: int,
+    drop_last: bool = True,
+):
+    """Unfold a timeseries into windows
+
+    Args:
+        x:
+            The timeseries to unfold. Can have shape
+            `(batch_size, num_channels, length * sample_rate)`,
+            `(num_channels, length * sample_rate)`, or
+            `(length * sample_rate)`
+        window_size:
+            The size of the windows to unfold from x
+        stride:
+            The stride between windows
+        drop_last:
+            If true, does not return the remainder that exists
+            when the timeseries cannot be evenly broken up into
+            windows
+
+    Returns:
+       A tensor of shape
+       `(num_windows, batch_size, num_channels, kernel_size)`,
+       `(num_windows, num_channels, kernel_size)`, or
+       `(num_windows, kernel_size)` depending on whether the
+       input tensor is 3D, 2D, or 1D
+
+       If `drop_last` is false, returns the remainder of the
+       timeseries, shaped to be compatible with the returned
+       unfolded tensor
+    """
+    reshape = list(x.shape[:-1])
+    if x.ndim == 1:
+        x = x[None, None, None, :]
+    elif x.ndim == 2:
+        x = x[None, :, None, :]
+    elif x.ndim == 3:
+        x = x[:, :, None, :]
+
+    num_windows = (x.shape[-1] - window_size) // stride + 1
+    stop = (num_windows - 1) * stride + window_size
+    x_clone = x.detach().clone()
```

Comment:
Is there a reason you need to detach from the graph and copy here?

Reply:
Honestly, I can't remember what I was thinking last week. I _think_ I wanted to have only one `if not drop_last` block, so I needed to copy `x` before it underwent its transformations, and I read that `.detach().clone()` was the preferred was to make a copy of a tensor. But that's not really necessary

Reply:
Ahhh ok gotcha. Yeah I think in this case you would just want to change the variable name for `x` once you started transforming it, because when you `detach` a tensor you stop all backwards gradient flow through that tensor (it emerges from the copy as if it was just born, with no relation to any transforms that might have brought the tensor from which it was cloned into being. If you've seen The Prestige, it's kind of like that). This means that things in the loss function which depend on this chunk won't be able to update the weights of transforms before it accordingly.